File: backend/app/models/luqun_repair.py
"""
樂群工務報修案件 ORM Model
對應 Ragic: ap12.ragic.com/soutlet001/luqun-public-works-repair-reporting-system/6

與 DazhiRepairCase 相比多了：
  - deduction_counter（扣款專櫃）
  - occ_year / occ_month（報修月份，4.1 統計用）
"""
from datetime import datetime
from app.core.time import twnow
from sqlalchemy import Boolean, DateTime, Float, Integer, String, Text
from sqlalchemy.orm import mapped_column, Mapped
from app.core.database import Base


class LuqunRepairCase(Base):
    __tablename__ = "luqun_repair_case"

    ragic_id:          Mapped[str]   = mapped_column(String(50),  primary_key=True)
    case_no:           Mapped[str]   = mapped_column(String(100), default="")
    title:             Mapped[str]   = mapped_column(Text,        default="")
    reporter_name:     Mapped[str]   = mapped_column(String(100), default="")
    repair_type:       Mapped[str]   = mapped_column(String(50),  default="")
    floor:             Mapped[str]   = mapped_column(String(100), default="")
    floor_normalized:  Mapped[str]   = mapped_column(String(30),  default="")
    occurred_at:       Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    responsible_unit:  Mapped[str]   = mapped_column(String(100), default="")
    work_hours:        Mapped[float] = mapped_column(Float,       default=0.0)
    status:            Mapped[str]   = mapped_column(String(50),  default="")
    outsource_fee:     Mapped[float] = mapped_column(Float,       default=0.0)
    maintenance_fee:   Mapped[float] = mapped_column(Float,       default=0.0)
    total_fee:         Mapped[float] = mapped_column(Float,       default=0.0)
    deduction_item:         Mapped[str]   = mapped_column(String(200), default="")
    deduction_fee:          Mapped[float] = mapped_column(Float,       default=0.0)
    deduction_counter:      Mapped[float] = mapped_column(Float,       default=0.0)  # 保持 0
    deduction_counter_name: Mapped[str]   = mapped_column(String(200), default="")   # 扣款專櫃名稱
    acceptor:               Mapped[str]   = mapped_column(String(100), default="")
    accept_status:          Mapped[str]   = mapped_column(String(200), default="")
    closer:                 Mapped[str]   = mapped_column(String(100), default="")
    finance_note:           Mapped[str]   = mapped_column(Text,        default="")
    mgmt_response:          Mapped[str]   = mapped_column(Text,        default="")   # 管理單位回應
    is_completed:      Mapped[bool]  = mapped_column(Boolean,     default=False)
    completed_at:      Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    close_days:        Mapped[float | None]    = mapped_column(Float, nullable=True)
    year:              Mapped[int | None]      = mapped_column(Integer, nullable=True)
    month:             Mapped[int | None]      = mapped_column(Integer, nullable=True)
    occ_year:          Mapped[int | None]      = mapped_column(Integer, nullable=True)   # 樂群專用
    occ_month:         Mapped[int | None]      = mapped_column(Integer, nullable=True)   # 樂群專用
    is_room_case:      Mapped[bool]  = mapped_column(Boolean,     default=False)
    room_no:           Mapped[str]   = mapped_column(String(20),  default="")
    room_category:     Mapped[str]   = mapped_column(String(50),  default="")
    synced_at:         Mapped[datetime] = mapped_column(DateTime, default=twnow)

    @property
    def is_completed_flag(self) -> bool:
        """有完工時間（completed_at）即視為已完工；亦兼顧狀態欄位判斷"""
        return (self.completed_at is not None) or self.is_completed

    def _get_counter_stores(self) -> list[str]:
        """計算扣款專櫃列表（供 compute_dashboard 使用）"""
        from app.services.luqun_repair_service import _parse_counter_stores
        return _parse_counter_stores(self.deduction_counter_name or "", self.finance_note or "")

    @property
    def counter_stores(self) -> list[str]:
        return self._get_counter_stores()

    @property
    def is_excluded_flag(self) -> bool:
        """RepairCase 相容屬性：排除「取消」等不計入統計的案件"""
        return self.status.strip() in {"取消"}

    def to_dict(self) -> dict:
        """轉換為 API 回傳用 dict（與 RepairCase.to_dict 介面相同）"""
        from datetime import datetime as _dt
        return {
            "ragic_id":               self.ragic_id,
            "case_no":                self.case_no,
            "title":                  self.title,
            "reporter_name":          self.reporter_name,
            "repair_type":            self.repair_type,
            "floor":                  self.floor,
            "floor_normalized":       self.floor_normalized,
            "occurred_at":            self.occurred_at.strftime("%Y/%m/%d %H:%M") if self.occurred_at else "",
            "responsible_unit":       self.responsible_unit,
            "work_hours":             self.work_hours,
            "status":                 self.status,
            "outsource_fee":          self.outsource_fee,
            "maintenance_fee":        self.maintenance_fee,
            "total_fee":              self.total_fee,
            "acceptor":               self.acceptor,
            "accept_status":          self.accept_status,
            "closer":                 self.closer,
            "deduction_item":         self.deduction_item,
            "deduction_fee":          self.deduction_fee,
            "deduction_counter":      self.deduction_counter,
            "deduction_counter_name": self.deduction_counter_name or "",
            "counter_stores":         self._get_counter_stores(),
            "mgmt_response":          self.mgmt_response or "",
            "finance_note":           self.finance_note,
            "is_completed":           self.is_completed_flag,
            "is_excluded":            self.is_excluded_flag,
            "completed_at":           self.completed_at.strftime("%Y/%m/%d") if self.completed_at else "",
            "close_days":             self.close_days,
            "pending_days": (
                round((_dt.now() - self.occurred_at).total_seconds() / 86400)
                if (self.completed_at is None and self.occurred_at)
                else None
            ),
            "year":               self.year,
            "month":              self.month,
            "occ_year":           self.occ_year,
            "occ_month":          self.occ_month,
            "is_room_case":       self.is_room_case,
            "room_no":            self.room_no,
            "room_category":      self.room_category,
            "images":             [],
        }

File: frontend/src/types/luqunRepair.ts
/**
 * 樂群工務報修 TypeScript 型別定義
 */

// ── 單筆報修案件 ──────────────────────────────────────────────────────────────
export interface RepairImage {
  url:      string
  filename: string
}

export interface RepairCase {
  ragic_id:         string
  case_no:          string
  title:            string
  reporter_name:    string
  repair_type:      string
  floor:            string
  floor_normalized: string
  occurred_at:      string
  responsible_unit: string
  work_hours:       number
  status:           string
  outsource_fee:    number
  maintenance_fee:  number
  total_fee:        number
  acceptor:         string
  accept_status:    string
  closer:           string
  deduction_item:         string
  deduction_fee:          number
  deduction_counter:      number        // 保持 0（欄位存名稱，非金額）
  deduction_counter_name: string        // 扣款專櫃名稱（如"牪肉舖"或"多櫃"）
  counter_stores:         string[]      // 解析後的專櫃列表
  mgmt_response:          string        // 管理單位回應
  finance_note:           string
  is_completed:     boolean
  completed_at:     string
  close_days:       number | null
  pending_days:     number | null
  year:             number | null
  month:            number | null
  is_room_case:     boolean
  room_no:          string
  room_category:    string
  images:           RepairImage[]
}

// ── Dashboard ─────────────────────────────────────────────────────────────────
export interface DashboardKpi {
  total:                    number
  completed:                number
  uncompleted:              number
  avg_close_days:           number | null
  total_fee:                number
  total_deduction_fee:      number
  total_deduction_counter:  number        // 保持 0
  total_counter_stores:     number        // 本月有扣款的專櫃家數
  total_counter_fee:        number        // 本月扣款專櫃費用合計
  counter_store_names:      string[]      // 本月專櫃名稱列表
  total_work_hours:         number
  room_cases:               number
  pending_verify:           number          // 待辦驗件數
  // 年度費用（費用 KPI 卡片用，不受月份篩選影響）
  annual_fee:                    number
  annual_outsource_fee:          number
  annual_maintenance_fee:        number
  annual_deduction_fee:          number
  annual_deduction_counter:      number
  // 年度扣款專櫃（全年統計，與費用 KPI 卡片對齊）
  annual_counter_stores:         number        // 全年有扣款的專櫃家數
  annual_counter_fee:            number        // 全年扣款專櫃費用合計
  annual_counter_store_names:    string[]      // 全年專櫃名稱列表
  // 當月費用（費用 KPI 卡片用，依年+月篩選）
  month_outsource_fee:     number
  month_maintenance_fee:   number
  month_deduction_fee:     number
  month_deduction_counter: number
  month_total_fee:         number
}

export interface Trend12MPoint {
  label:     string
  year:      number
  month:     number
  total:     number
  completed: number
}

export interface TypeDistItem {
  type:  string
  count: number
}

export interface FloorDistItem {
  floor: string
  count: number
}

export interface StatusDistItem {
  status: string
  count:  number
}

export interface DashboardData {
  kpi:               DashboardKpi
  trend_12m:         Trend12MPoint[]
  type_dist:         TypeDistItem[]
  floor_dist:        FloorDistItem[]
  status_dist:       StatusDistItem[]
  top_uncompleted:   RepairCase[]
  top_fee:           RepairCase[]
  top_hours:         RepairCase[]
  // KPI 明細（點擊 KPI 卡片時用）
  kpi_total_detail:            RepairCase[]
  kpi_completed_detail:        RepairCase[]
  kpi_uncompleted_detail:      RepairCase[]
  kpi_pending_verify_detail:   RepairCase[]
  kpi_close_days_detail:       RepairCase[]
  kpi_room_detail:             RepairCase[]
  kpi_hours_detail:            RepairCase[]
  kpi_counter_stores_detail:   RepairCase[]
  // 年度費用明細（點擊費用卡片時用）
  annual_fee_detail:       RepairCase[]
  annual_deduction_detail: RepairCase[]
  annual_counter_detail:   RepairCase[]
}

// ── 4.1 報修統計 ──────────────────────────────────────────────────────────────
export interface MonthRepairStat {
  month:                      number
  prev_uncompleted:           number
  closed_from_prev:           number
  prev_remaining:             number
  cum_completion_rate:        number | null
  this_month_total:           number
  this_month_completed:       number
  this_month_uncompleted:     number
  this_month_completion_rate: number | null
  // 明細（點擊展開）
  prev_uncompleted_detail:    RepairCase[]
  closed_from_prev_detail:    RepairCase[]
  prev_remaining_detail:          RepairCase[]
  this_month_uncompleted_detail:  RepairCase[]
  this_month_total_detail:    RepairCase[]
  this_month_completed_detail:RepairCase[]
}

export interface RepairStatsData {
  year:   number
  months: Record<number, MonthRepairStat>
}

// ── 4.2 結案時間統計 ──────────────────────────────────────────────────────────
export interface ClosingBlock {
  closed_count: number
  total_days:   number
  avg_days:     number | null
  cases:        RepairCase[]
}

export interface ClosingMonthlyItem {
  small: ClosingBlock
  large: ClosingBlock
}

export interface ClosingTimeData {
  year:                  number
  month:                 number | null
  small:                 ClosingBlock
  large:                 ClosingBlock
  monthly:               Record<number, ClosingMonthlyItem>
  classification_note:   string
}

// ── 4.3 報修類型統計 ──────────────────────────────────────────────────────────
export interface TypeStatRow {
  type:           string
  example:        string
  monthly:        Record<number, number>
  monthly_detail: Record<number, RepairCase[]>
  row_total:      number
  prev_month:     number
  this_month:     number
  cum_pct:        number
}

export interface TypeStatsData {
  year:        number
  focus_month: number | null
  rows:        TypeStatRow[]
  year_total:  number
  type_order:  string[]
}

// ── 4.4 客房報修表 ────────────────────────────────────────────────────────────
export interface RoomCategoryEntry {
  ragic_id: string
  title:    string
  status:   string
}

export interface RoomRepairRow {
  room_no:    string
  floor:      string
  categories: Record<string, RoomCategoryEntry[]>
}

export interface RoomRepairTableData {
  year:               number
  month:              number
  categories:         string[]
  rows:               RoomRepairRow[]
  unknown_room_cases: RepairCase[]
  floors_with_data:   string[]
  total_room_cases:   number
}

// ── 明細查詢 ──────────────────────────────────────────────────────────────────
export interface DetailQueryParams {
  year?:        number
  month?:       number
  repair_type?: string
  floor?:       string
  status?:      string
  keyword?:     string
  page?:        number
  page_size?:   number
  sort_by?:     string
  sort_desc?:   boolean
}

export interface DetailResult {
  total:     number
  page:      number
  page_size: number
  items:     RepairCase[]
}

// ── 過濾選項 ──────────────────────────────────────────────────────────────────
export interface FilterOptions {
  repair_types: string[]
  floors:       string[]
  statuses:     string[]
}

// ── 金額統計 ──────────────────────────────────────────────────────────────────
export type FeeKey = 'outsource_fee' | 'maintenance_fee' | 'deduction_fee' | 'deduction_counter'

export interface FeeMonthRow {
  outsource_fee:     number
  maintenance_fee:   number
  deduction_fee:     number
  deduction_counter: number
}

export interface FeeTotals {
  outsource_fee:     number
  maintenance_fee:   number
  deduction_fee:     number
  deduction_counter: number
}

export interface FeeStatsData {
  year:            number
  monthly_totals:  Record<number, FeeMonthRow>   // key = 1..12
  fee_totals:      FeeTotals                      // annual total per fee type
  month_totals:    Record<number, number>          // monthly sum of all fee types
  grand_total:     number
  monthly_detail:  Record<string, RepairCase[]>   // key = "{m}_{fk}"
  fee_labels:      Record<FeeKey, string>
}
